src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/Debugger/WebServer.cs

```diff
@@ -0,0 +1,1022 @@
+
+        /// <summary>
+        /// Hooks into IIS events to detect when the applicationhost.config file changes. This can be called many times. It is
+        /// ignored if already listening
+        /// </summary>
+        private void AdviseIISServiceEvents()
+        {
+            if (_iisServiceEventsCookie == 0)
+            {
+                LocalIISService3.AdviseIISServiceEvents(this, out _iisServiceEventsCookie);
+            }
+        }
+
+        /// <summary>
+        /// Unhooks from IIS service events
+        /// </summary>
+        private void UnadviseIISServiceEvents()
+        {
+            if (_iisServiceEventsCookie != 0)
+            {
+                LocalIISService3.UnadviseIISServiceEvents(_iisServiceEventsCookie);
+                _iisServiceEventsCookie = 0;
+            }
+        }
+
+        /// <summary>
+        /// Called just prior to the applicationhost.config file path changing. We must close any web server, so
+        /// that it forces a re-get of these on the next run.
+        /// </summary>
+        public void BeforeMoveApplicationHostConfigFile()
+        {
+            ReleaseVsDeveloperWebServer();
+        }
+
+        /// <summary>
+        /// Note that in a couple of scenarios, the Before function is not called first so we
+        /// also run the cleanup here.
+        /// </summary>
+        public void AfterMoveApplicationHostConfigFile()
+        {
+            ReleaseVsDeveloperWebServer();
+        }
+
+        private string GetErrorInfo()
+        {
+            _uiShell.Value.GetErrorInfo(out string errText);
+
+            // Sometimes the shell returns trailing nulls in the string. These need to be removed
+            return errText == null ? string.Empty : errText.TrimEnd('\0');
+        }
+
+        private void SetErrorInfo(int hr, string description)
+        {
+            _uiShell.Value.SetErrorInfo(hr, description, 0, null, string.Empty);
+        }
+
+        private bool VSIsRunningElevated()
+        {
+            IVsShell3 vsShell = (IVsShell3)_vsShell.Value;
+
+            vsShell.IsRunningElevated(out bool bIsElevated);
+            return bIsElevated;
+        }
+        private static bool DirectoriesMatch(string dir1, string dir2)
+        {
+            return string.Compare(EnsureTrailingChar(dir1, '\\'), EnsureTrailingChar(dir2, '\\'), StringComparison.OrdinalIgnoreCase) == 0;
+        }
+
+        public static bool DirectoryIsEmpty(string folderPath)
+        {
+            if (Directory.Exists(folderPath))
+            {
+                if (Directory.EnumerateFiles(folderPath).Any())
+                {
+                    return false;
+                }
+
+                foreach (string dir in Directory.EnumerateDirectories(folderPath, "*.*", SearchOption.TopDirectoryOnly))
+                {
+                    if (dir != "." && dir != "..")
+                    {
+                        return false;
+                    }
+                }
```

Comment:
Can this be acheived in a single enumeration via `Directory.EnumerateFileSystemEntries`?

Reply:
Good point